### PR TITLE
Requires with error message

### PR DESCRIPTION
**Pull request by bazsi**
This is a continuation of #2979 to improve error messages if an installed SCL is missing a dependency.

With these changes we are going to have this as an error message if we are using elasticsearch-http() without the http() module present:

```
[2019-11-14T21:18:09.766325] Cannot load required module; module='http', details='The http() driver is required for elasticsearch-http(), please install syslog-ng-mod-http (Debian) or syslog-ng-http (RHEL) package', location='etc/syslog-ng-elastic.conf:7:16'
Error parsing log statement, syntax error, unexpected LL_ERROR, expecting '}' in block destination elasticsearch-http() at /install/share/syslog-ng/include/scl/elasticsearch/elastic-http.conf:24:4:1-5:1:
1       
2       #Start Block block destination elasticsearch-http() at /install/share/syslog-ng/include/scl/elasticsearch/elastic-http.conf:24
3       
4-----> @requires http "The http() driver is required for elasticsearch-http(), please install syslog-ng-mod-http (Debian) or syslog-ng-http (RHEL) package"
4-----> ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
5----->   http(
6         url("http://localhost:1234/")
7         headers("Content-Type: application/x-ndjson")
8         workers(4)
9         batch_lines(100)

Included from etc/syslog-ng-elastic.conf:7:16-7:97:
2       
3       @include "scl/elasticsearch/elastic-http.conf"
4       
5       log {
6       	source { tcp(port(2000)); };
7-----> 	destination { elasticsearch-http(index("syslog") type("syslog") url("http://localhost:1234/")); };
7-----> 	              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
8       };


```

### CODE REVIEW

**Comment by kira-syslogng** [comment]
Build FAILURE

**Comment by kira-syslogng** [comment]
Build FAILURE

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Comment by kira-syslogng** [comment]
Build SUCCESS

File: scl/slack/slack.conf [hunk 1]
```diff
@@ -21,8 +21,6 @@
 #############################################################################
 
 @requires http
-@requires json-plugin
-@requires basicfuncs
```
Comment (Kokan) [inline scl/slack/slack.conf]: Please do not remove this; either move them into the block like http or leave them here.

File: scl/slack/slack.conf [hunk 2]
```diff
@@ -45,6 +43,9 @@ block destination slack(
   use-system-cert-store(yes)
   ...)
 {
+
+@requires http "The slack() driver depends on the syslog-ng http module, please install the syslog-ng-mod-http (Debian & derivatives) or the syslog-ng-http (RHEL & co) package"
```
Comment (Kokan) [inline scl/slack/slack.conf]: The `@requires http` left on top of the file that defeats a purpose of this, please remove that.

File: scl/elasticsearch/plugin.conf [hunk 3]
```diff
@@ -20,7 +20,6 @@
 #
 #############################################################################
 
-@requires json-plugin
```
Comment (Kokan) [inline scl/elasticsearch/plugin.conf]: Please do not remove this, either keep it here or push down into the block.

File: scl/scl.conf [hunk 4]
```diff
@@ -25,6 +25,16 @@
 # `include-path`, then includes all SCL supplied plugins.
 #
 
-@include 'scl/*/*.conf'
-
 @define java-module-dir "`module-install-dir`/java-modules"
+
+@requires json-plugin \
```
Comment (Kokan) [inline scl/scl.conf]: The `@require` initially have been done due to json plugin not available all the time. (https://github.com/syslog-ng/syslog-ng/issues/673).

I would first of all split this part, as the requires message work is fine. This part where you make json mandatory is kinda a different issue. (If we want to do that I would also make it not optional to compile json module, as in this way it is really easy to make it fail.)

**Comment by bazsi** [comment]
Re: hardwiring json-plugin dependency

I think it is entirely different on what pure syslog-ng depends on and what the SCL supplied modules depend on. 

JSON is pretty much a hard dependency and has been included in the "core" packages on both Debian and RedHat packaging. With a closer check I can see that it is made part of the default package everywhere.

Debian: https://github.com/syslog-ng/syslog-ng/blob/5e4d0f2f759e80d0ad1f6c4f48a294185d5a9994/packaging/debian/syslog-ng-core.install#L31
RedHat: https://github.com/syslog-ng/syslog-ng/blob/5e4d0f2f759e80d0ad1f6c4f48a294185d5a9994/packaging/rhel/syslog-ng.spec#L472
FreeBSD: https://svnweb.freebsd.org/ports/head/sysutils/syslog-ng324/Makefile?view=markup&pathrev=519926#l33
Arch: https://www.archlinux.org/packages/extra/x86_64/syslog-ng/

I think it is ok that we enforce JSON dependency for SCL. If someone does not want JSON he can choose not to use the SCL either. And the fragmentation of SCL that is present in some of the packagings is exactly the reason why SCL is far from an ideal state and that cannot be depended upon. Thus my aim to simplify things and make sure the "default behavior" is really a default that can be depended on.

With regards to build options, json-c has been integrated as a submodule exactly because our dependency on it is so large (and has been increasing steadily). On my part it would be ok to make it a hard dependency.

File: scl/slack/slack.conf (hunk 2 above)
Comment (bazsi) [reply scl/slack/slack.conf]: done.

File: scl/slack/slack.conf (hunk 1 above)
Comment (bazsi) [reply scl/slack/slack.conf]: I have separated this to the last patch, which can be dropped depending on the discussion we have on making json-plugin mandatory. But I'm resolving this conversation.

File: scl/elasticsearch/plugin.conf (hunk 3 above)
Comment (bazsi) [reply scl/elasticsearch/plugin.conf]: again, this one is pending on the discussion on making json-plugin mandatory. resolving this.

File: scl/scl.conf (hunk 4 above)
Comment (bazsi) [reply scl/scl.conf]: I have now moved all of this into a separate patch that can be split off to a separate branch if needed, I just didn't want to add the detailed comment on the missing module into n+1 locations.

like I said in the other thread, at this point I would opt for making json-plugin mandatory for the SCL. IF someone doesn't need the functionality of the SCL they can forgo it simply by removing the "@include scl.conf" in their configuration.

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Comment by bazsi** [comment]
I've just dropped the last patch of the branch as we agreed IRL. I am opening a separate PR on making JSON mandatory.

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Pull request merged**